### ISSUE TITLE
Non-optimal pedestrian routes on slopes when going down

### ISSUE THREAD

**Issue by szymone** [valhalla/valhalla#3550]
There is an issue with finding optimal routes using pedestrian routing on slopes when going down. Found routes can have higher cost than the optimal one. I assume this is caused by the heuristic factor for the A* algorithm and the way how it is currently calculated.

When you are going down the slope the speed factor can be multiplied by 0.67 based on kGradeBasedSpeedFactor[0] value, which is less than 1.0. This is calculated in EdgeCost() method. https://github.com/valhalla/valhalla/blob/66c28863e6bbf785b0c346ed72b2559231cd22cd/src/sif/pedestriancost.cc#L696 

However in AStarCostFactor() the 0.67 value is not used. https://github.com/valhalla/valhalla/blob/66c28863e6bbf785b0c346ed72b2559231cd22cd/src/sif/pedestriancost.cc#L398

Changing the line `return (speedfactor_ * factor);` to `return (0.67f * speedfactor_ * factor);` in AStarCostFactor() resolves the issue and the resulted route is optimal. All other factors are equal or grater than 1.0.

Can anyone confirm this is a correct way of fixing this?

Also this can be related to #3547.

**Comment by kevinkreiser** [comment]
cc @dnesbitt61

**Comment by dnesbitt61** [comment]
The ACostFactor method should account for favorable weights applied and use those to reduce the A* factor being used. This will reduce performance unfortunately. Since the favorable weighting for downhill sections essentially reduces the cost (time) along these downhill edges - the A* factor must be reduced as well. If not, then it may overestimate the actual cost (time) to get from the current location in the graph search to the destination - making the heuristic "inadmissible".

**Comment by kevinkreiser** [comment]
yeah we should make a constant out of the minimum factor due to down hill grades and update the heuristic to under estimate that goal again. in fact the constant exists as the first entry here: https://github.com/valhalla/valhalla/pull/3234/files#diff-c7c3e5a11f81e05708af6a935632985ffa39c93f66d330578565ac72c19f375eR165-R182

we should probably also revalidate the numbers in that array to make sure they match the function they reference

**Comment by szymone** [comment]
@dnesbitt61, @kevinkreiser thanks for your comments. I can create PR for this, however I have two more questions related to this issue. 

1. There is an avoid_hills factor based on use_hills costing option. Currently it is only used for weighting the grade_penalty, should it be used to weight kGradeBasedSpeedFactor as well?
2. What about sac scale factor which can lower the speed and increase the additional penalty factors in costing? I think sac scale for hiking trails is also based on grade. If so, then the grade is used in speed and penalty factor calculations twice. Once directly using grade penalty and grade based speed factor and second, indirectly by applying the sac scale. Is this true? I think we could add use_sac_scale costing option to have more control over speed and penalty factors calculations based on sac scale. What do you think?

**Comment by dnesbitt61** [comment]
1. I don't think so. That would impact estimated time for the route, which should not be impacted by costing penalties.
2. Perhaps if sac_scale is set on an edge, it should be used to modulate speed (override any grade factor) rather than combine with the weighted grade. But I haven't experimented with this. I agree that penalizing an edge twice doesn't seem good.
